interactors/src/main/kotlin/interactors/StudentsInteractor.kt

```diff
@@ -0,0 +1,22 @@
+package com.students.results.interactors
+
+import arrow.core.Either
+import arrow.core.flatMap
+import com.students.results.repository.ExamsRepository
+import com.students.results.repository.StudentsRepository
+import com.students.results.services.NotateExamException
+import com.students.results.services.Students
+import com.students.results.services.requests.NotateExam
+
+class StudentsInteractor(private val studentsRepository: StudentsRepository, private val examsRepository: ExamsRepository) : Students {
+
+    override fun notate(notateExam: NotateExam): Either<NotateExamException, Unit> =
+            studentsRepository.findStudentById(studentId = 1).flatMap { student ->
+                examsRepository.findExamById(examId = notateExam.examId).flatMap { exam ->
+                    student.notate(exam, notateExam.note).flatMap { updatedStudent ->
+                        studentsRepository.save(updatedStudent)
```

Comment:
There are too many levels of abstraction here and it is not clear what this method is doing. Remember the SRP :)

Reply:
What would you remove there is only two call and entity method.
Maybe you do not like flatMap with either, but I don't see how to do it shorter than that

Reply:
Maybe by giving up either, but I want to use it to handle errors

Reply:
Mostly that we jump from a student into an exam into a grade for that exam and student. all the wile nesting.

Reply:
Nesting helps to get the values directly. The other solution is to do if(isRight()=> toOption().get()) etc..etc but I really don't like that solution (I tested it)

Reply:
I searched a lot, and normally the following code is better and compile (I need to check at home)
```kotlin
notateExam.also {
    ForEither<RepositoryException>() extensions {
        tupled(
            studentsRepository.findStudentById(studentId),
            examsRepository.findExamById(examId)
        )
    }.flatMap { (student, exam) ->
        student.notate(exam, note).flatMap { studentsRepository.save(it) }
    }.mapLeft { NotateExamException(throwable = it) }
}
 ```

Reply:
It does not work, please close it, it is useless